docs/orleans/grains/grain-extensions.md

```diff
@@ -0,0 +1,160 @@
+
+```csharp
+var grain = client.GetGrain<SomeExampleGrain>(someKey);
+var grainReferenceAsInterface = grain.AsReference<IGrainDeactivateExtension>();
+
+await grainReferenceAsInterface.Deactivate("Because, I said so...");
+```
+
+## State manipulation extension example
+
+In this example, you learn how to add a `GetState` and `SetState` method to any grain through extensions, allowing you to manipulate the grain's internal state.
+
+### State manipulation extension interface
+
+First, define the `IGrainStateAccessor<T>` interface, which contains the `GetState` and `SetState` methods. Again, this interface must derive from `IGrainExtension`.
+
+```csharp
+public interface IGrainStateAccessor<T> : IGrainExtension
+{
+    Task<T> GetState();
+    Task SetState(T state);
+}
+```
+
+Once you have access to the target grain, you can use the extension to manipulate its state. In this example, you use an extension to access and modify a specific integer state value within the target grain.
+
+### State manipulation extension implementation
+
+The extension you use is `IGrainStateAccessor<T>`, which provides methods to get and set a state value of type `T`. To create the extension, you implement the interface in a class that takes a `getter` and a `setter` as arguments in its constructor.
+
+```csharp
+public sealed class GrainStateAccessor<T> : IGrainStateAccessor<T>
+{
+    private readonly Func<T> _getter;
+    private readonly Action<T> _setter;
+
+    public GrainStateAccessor(Func<T> getter, Action<T> setter)
+    {
+        _getter = getter;
+        _setter = setter;
+    }
+
+    public Task<T> GetState()
+    {
+        return Task.FromResult(_getter.Invoke());
+    }
+
+    public Task SetState(T state)
+    {
+        _setter.Invoke(state);
+        return Task.CompletedTask;
+    }
+}
+```
+
+In the preceding implementation, the `GrainStateAccessor<T>` class takes `getter` and `setter` arguments in its constructor. These delegates are used to read and modify the target grain's state. The `GetState()` method returns a <xref:System.Threading.Tasks.Task%601> wraps the current value of the `T` state, while the `SetState(T state)` method sets the new value of the `T` state.
+
+### State manipulation extension registration and usage
+
+To use the extension to access and modify the target grain's state, you need to register the extension and set its components in the <xref:Orleans.Grain.OnActivateAsync?displayProperty=nameWithType> method of the target grain.
+
+```csharp
+public override Task OnActivateAsync()
+{
+    // Retrieve the IGrainStateAccessor<T> extension
+    var accessor = new GrainStateAccessor<int>(
+        getter: () => this.Value,
+        setter: value => this.Value = value);
+
+    // Set the extension as a component of the target grain's context
+    var context = _contextAccessor.GrainContext;
+    context.SetComponent<IGrainStateAccessor<int>>(accessor);
```

Comment:
```suggestion
    ((IGrainBase)this).GrainContext.SetComponent<IGrainStateAccessor<int>>(accessor);
```

Since we're not showing where `IGrainContextAccessor` or `IGrainContext` is injected here, maybe we should use a portable approach like this.